File: ios/Sources/MultandoSDK/Core/HTTPClient.swift
import Foundation

/// Async HTTP client backed by URLSession with automatic auth header injection,
/// 401 token refresh, and snake_case JSON coding.
public final class HTTPClient: @unchecked Sendable {

    // MARK: - Properties

    private let session: URLSession
    private let config: MultandoConfig
    private let authManager: AuthManager
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    /// Lock used to coalesce concurrent 401 refresh attempts.
    private let refreshLock = NSLock()
    private var isRefreshing = false

    // MARK: - Init

    init(config: MultandoConfig, authManager: AuthManager) {
        self.config = config
        self.authManager = authManager

        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = config.timeout
        self.session = URLSession(configuration: sessionConfig)

        let enc = JSONEncoder()
        enc.keyEncodingStrategy = .convertToSnakeCase
        enc.dateEncodingStrategy = .iso8601
        self.encoder = enc

        let dec = JSONDecoder()
        dec.keyDecodingStrategy = .convertFromSnakeCase
        dec.dateDecodingStrategy = .iso8601
        self.decoder = dec
    }

    // MARK: - Public API

    func request<T: Decodable>(
        method: String,
        path: String,
        body: (any Encodable)? = nil,
        queryItems: [URLQueryItem]? = nil,
        authenticated: Bool = true
    ) async throws -> T {
        let data = try await raw(method: method, path: path, body: body, queryItems: queryItems, authenticated: authenticated)
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw MultandoError.decodingError(error)
        }
    }

    func requestVoid(
        method: String,
        path: String,
        body: (any Encodable)? = nil,
        queryItems: [URLQueryItem]? = nil,
        authenticated: Bool = true
    ) async throws {
        _ = try await raw(method: method, path: path, body: body, queryItems: queryItems, authenticated: authenticated)
    }

    // MARK: - Internal

    private func raw(
        method: String,
        path: String,
        body: (any Encodable)?,
        queryItems: [URLQueryItem]?,
        authenticated: Bool,
        isRetry: Bool = false
    ) async throws -> Data {
        var urlRequest = try buildRequest(method: method, path: path, body: body, queryItems: queryItems, authenticated: authenticated)

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch {
            throw MultandoError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw MultandoError.networkError(URLError(.badServerResponse))
        }

        // 401 → attempt token refresh once
        if httpResponse.statusCode == 401 && authenticated && !isRetry {
            try await refreshToken()
            return try await raw(method: method, path: path, body: body, queryItems: queryItems, authenticated: authenticated, isRetry: true)
        }

        // 429 → try to decode the backend's structured rate-limit body.
        if httpResponse.statusCode == 429 {
            if let typed = Self.decodeRateLimitError(data: data, response: httpResponse, body: body) {
                throw typed
            }
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            let message = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw MultandoError.apiError(statusCode: httpResponse.statusCode, message: message)
        }

        return data
    }

    // MARK: - Rate-limit decoding

    /// FastAPI wraps HTTPException(detail=...) as `{"detail": {...}}`.
    private struct RateLimitEnvelope: Decodable {
        let detail: RateLimitDetail
    }

    private struct RateLimitDetail: Decodable {
        let error: String?
        let limit: String?
        let retryAfterSeconds: Int?
        let windowSeconds: Int?

        enum CodingKeys: String, CodingKey {
            case error
            case limit
            case retryAfterSeconds = "retry_after_seconds"
            case windowSeconds = "window_seconds"
        }
    }

    /// Maps a 429 response body + headers onto our typed error cases.
    /// Returns `nil` when the payload doesn't match the known schema so the
    /// caller falls back to a generic ``MultandoError.apiError``.
    private static func decodeRateLimitError(
        data: Data,
        response: HTTPURLResponse,
        body: (any Encodable)?
    ) -> MultandoError? {
        let dec = JSONDecoder()
        let detail: RateLimitDetail?
        if let envelope = try? dec.decode(RateLimitEnvelope.self, from: data) {
            detail = envelope.detail
        } else if let flat = try? dec.decode(RateLimitDetail.self, from: data) {
            detail = flat
        } else {
            detail = nil
        }

        // Retry-After header takes precedence; fall back to the body.
        let headerRetry = response.value(forHTTPHeaderField: "Retry-After").flatMap(Int.init)
        let retryAfter = TimeInterval(headerRetry ?? detail?.retryAfterSeconds ?? 0)

        switch detail?.limit {
        case "reports_per_hour":
            return .rateLimitExceeded(retryAfter: retryAfter, scope: .hour)
        case "reports_per_day":
            return .rateLimitExceeded(retryAfter: retryAfter, scope: .day)
        case "same_plate_per_user_24h", "plate_reports_24h":
            let plate = Self.extractPlate(from: body) ?? ""
            let hours = max(1, Int((detail?.windowSeconds ?? 0) / 3600))
            return .plateCooldown(plate: plate, retryAfterHours: hours)
        default:
            return nil
        }
    }

    /// Best-effort extraction of `license_plate` from an outgoing report body.
    private static func extractPlate(from body: (any Encodable)?) -> String? {
        guard let body else { return nil }
        let enc = JSONEncoder()
        enc.keyEncodingStrategy = .convertToSnakeCase
        guard
            let data = try? enc.encode(AnyEncodable(body)),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return json["license_plate"] as? String
    }

    private func buildRequest(
        method: String,
        path: String,
        body: (any Encodable)?,
        queryItems: [URLQueryItem]?,
        authenticated: Bool
    ) throws -> URLRequest {
        guard var components = URLComponents(string: "\(config.baseURL)\(path)") else {
            throw MultandoError.validationError("Invalid URL: \(config.baseURL)\(path)")
        }
        if let queryItems, !queryItems.isEmpty {
            components.queryItems = queryItems
        }
        guard let url = components.url else {
            throw MultandoError.validationError("Could not construct URL from components")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(config.locale, forHTTPHeaderField: "Accept-Language")
        request.setValue(config.apiKey, forHTTPHeaderField: "X-API-Key")

        if authenticated, let token = authManager.accessToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            request.httpBody = try encoder.encode(AnyEncodable(body))
        }

        return request
    }

    private func refreshToken() async throws {
        guard let refreshToken = authManager.refreshToken else {
            throw MultandoError.authError("No refresh token available")
        }

        let body = RefreshRequest(refreshToken: refreshToken)
        let tokenResponse: TokenResponse = try await request(
            method: "POST",
            path: "/api/v1/auth/refresh",
            body: body,
            authenticated: false
        )
        authManager.store(tokens: tokenResponse)
    }
}

// MARK: - AnyEncodable helper

private struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void

    init(_ value: any Encodable) {
        _encode = { encoder in
            try value.encode(to: encoder)
        }
    }

    func encode(to encoder: Encoder) throws {
        try _encode(encoder)
    }
}

File: android/multando-sdk/src/main/kotlin/com/multando/sdk/core/HttpClient.kt
package com.multando.sdk.core

import com.multando.sdk.models.MultandoError
import com.multando.sdk.models.RateLimitScope
import com.multando.sdk.models.RefreshRequest
import com.multando.sdk.models.TokenResponse
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.KSerializer
import kotlinx.serialization.serializer
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import retrofit2.Retrofit
import java.util.concurrent.TimeUnit

/**
 * HTTP client built on OkHttp with automatic auth header injection,
 * 401 token refresh via an Authenticator, and snake_case JSON conversion.
 */
class HttpClient internal constructor(
    private val config: MultandoConfig,
    private val authManager: AuthManager
) {
    internal val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        isLenient = true
        coerceInputValues = true
    }

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    private val authInterceptor = Interceptor { chain ->
        val original = chain.request()
        val builder = original.newBuilder()
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .header("Accept-Language", config.locale)
            .header("X-API-Key", config.apiKey)

        if (original.header("No-Auth") == null) {
            authManager.accessToken?.let {
                builder.header("Authorization", "Bearer $it")
            }
        } else {
            builder.removeHeader("No-Auth")
        }

        chain.proceed(builder.build())
    }

    private val tokenAuthenticator = object : Authenticator {
        override fun authenticate(route: Route?, response: Response): Request? {
            // Only retry once
            if (response.request.header("X-Retry-Auth") != null) return null

            val refreshToken = authManager.refreshToken ?: return null
            val tokenResponse = try {
                runBlocking {
                    refreshTokens(refreshToken)
                }
            } catch (e: Exception) {
                authManager.clearTokens()
                return null
            }

            authManager.store(tokenResponse)

            return response.request.newBuilder()
                .header("Authorization", "Bearer ${tokenResponse.accessToken}")
                .header("X-Retry-Auth", "true")
                .build()
        }
    }

    internal val okHttpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(config.timeoutSeconds, TimeUnit.SECONDS)
        .readTimeout(config.timeoutSeconds, TimeUnit.SECONDS)
        .writeTimeout(config.timeoutSeconds, TimeUnit.SECONDS)
        .addInterceptor(authInterceptor)
        .authenticator(tokenAuthenticator)
        .build()

    // MARK: - Public request methods

    suspend inline fun <reified T> request(
        method: String,
        path: String,
        body: Any? = null,
        queryParams: Map<String, String>? = null,
        authenticated: Boolean = true
    ): T = withContext(Dispatchers.IO) {
        val responseBody = rawRequest(method, path, body, queryParams, authenticated)
        try {
            json.decodeFromString<T>(responseBody)
        } catch (e: Exception) {
            throw MultandoError.DecodingError(e)
        }
    }

    suspend fun requestVoid(
        method: String,
        path: String,
        body: Any? = null,
        queryParams: Map<String, String>? = null,
        authenticated: Boolean = true
    ) = withContext(Dispatchers.IO) {
        rawRequest(method, path, body, queryParams, authenticated)
    }

    // MARK: - Internal

    @PublishedApi
    internal suspend fun rawRequest(
        method: String,
        path: String,
        body: Any? = null,
        queryParams: Map<String, String>? = null,
        authenticated: Boolean = true
    ): String = withContext(Dispatchers.IO) {
        val urlBuilder = okhttp3.HttpUrl.Builder()
            .scheme(if (config.baseUrl.startsWith("https")) "https" else "http")

        val baseWithoutScheme = config.baseUrl
            .removePrefix("https://")
            .removePrefix("http://")

        val hostParts = baseWithoutScheme.split("/", limit = 2)
        val hostAndPort = hostParts[0]
        val basePath = if (hostParts.size > 1) "/" + hostParts[1] else ""

        if (hostAndPort.contains(":")) {
            val (host, port) = hostAndPort.split(":")
            urlBuilder.host(host).port(port.toInt())
        } else {
            urlBuilder.host(hostAndPort)
        }

        val fullPath = "$basePath$path".trimStart('/')
        fullPath.split("/").forEach { segment ->
            if (segment.isNotEmpty()) urlBuilder.addPathSegment(segment)
        }

        queryParams?.forEach { (key, value) ->
            urlBuilder.addQueryParameter(key, value)
        }

        val url = urlBuilder.build()

        val requestBody = when {
            body == null && method in listOf("POST", "PUT", "PATCH") ->
                "{}".toRequestBody(jsonMediaType)
            body is String ->
                body.toRequestBody(jsonMediaType)
            body != null ->
                json.encodeToString(serializer(body::class.java), body).toRequestBody(jsonMediaType)
            else -> null
        }

        val requestBuilder = Request.Builder()
            .url(url)
            .method(method, requestBody)

        if (!authenticated) {
            requestBuilder.header("No-Auth", "true")
        }

        val request = requestBuilder.build()
        val response = okHttpClient.newCall(request).execute()

        val responseBodyString = response.body?.string() ?: ""

        if (!response.isSuccessful) {
            if (response.code == 429) {
                mapRateLimitError(responseBodyString, response.header("Retry-After"))?.let { throw it }
            }
            throw MultandoError.ApiError(
                statusCode = response.code,
                message = responseBodyString.ifEmpty { response.message }
            )
        }

        responseBodyString
    }

    /**
     * Parse a structured 429 body into a typed [MultandoError.RateLimitException] or
     * [MultandoError.PlateCooldownException]. Returns null if the body isn't
     * recognized — the caller will fall back to a generic [MultandoError.ApiError].
     *
     * Expected body shape:
     * ```
     * { "detail": {
     *     "error": "rate_limit_exceeded",
     *     "limit": "reports_per_hour" | "reports_per_day" | "same_plate_per_user_24h" | "plate_reports_24h",
     *     "retry_after_seconds": 3600,
     *     "message": "..."
     *   }
     * }
     * ```
     * The SDK also accepts legacy bodies where the payload is flat (no
     * ``detail`` wrapper) or uses ``error_code`` instead of ``error``.
     */
    private fun mapRateLimitError(body: String, retryAfterHeader: String?): MultandoError? {
        if (body.isBlank()) return null
        val root = try {
            json.parseToJsonElement(body).jsonObject
        } catch (_: Exception) {
            return null
        }

        // FastAPI wraps HTTPException payloads under "detail".
        val payload: JsonObject = (root["detail"] as? JsonObject) ?: root

        val code = payload["error"]?.jsonPrimitive?.contentOrNull
            ?: payload["error_code"]?.jsonPrimitive?.contentOrNull
        val limit = payload["limit"]?.jsonPrimitive?.contentOrNull
        val message = payload["message"]?.jsonPrimitive?.contentOrNull
            ?: body

        val retryAfterSeconds: Long = payload["retry_after_seconds"]?.jsonPrimitive?.longOrNull
            ?: retryAfterHeader?.toLongOrNull()
            ?: 0L

        val isPlateCooldown = code == "plate_cooldown" ||
            limit == "same_plate_per_user_24h" ||
            limit == "plate_reports_24h"

        if (isPlateCooldown) {
            val plate = payload["plate"]?.jsonPrimitive?.contentOrNull ?: ""
            val retryAfterHours = ((retryAfterSeconds + 3599) / 3600).toInt().coerceAtLeast(1)
            return MultandoError.PlateCooldownException(
                message = message,
                plate = plate,
                retryAfterHours = retryAfterHours,
            )
        }

        if (code == "rate_limit_exceeded") {
            val scope = when (limit) {
                "reports_per_day" -> RateLimitScope.DAY
                else -> RateLimitScope.HOUR
            }
            return MultandoError.RateLimitException(
                message = message,
                retryAfterSeconds = retryAfterSeconds,
                scope = scope,
            )
        }

        return null
    }

    private suspend fun refreshTokens(refreshToken: String): TokenResponse {
        val body = json.encodeToString(RefreshRequest.serializer(), RefreshRequest(refreshToken))
        val requestBody = body.toRequestBody(jsonMediaType)

        val url = "${config.baseUrl}/api/v1/auth/refresh"
        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .header("Content-Type", "application/json")
            .header("X-API-Key", config.apiKey)
            .build()

        return withContext(Dispatchers.IO) {
            val response = okHttpClient.newCall(request).execute()
            val responseBody = response.body?.string() ?: ""
            if (!response.isSuccessful) {
                throw MultandoError.ApiError(response.code, responseBody)
            }
            json.decodeFromString(TokenResponse.serializer(), responseBody)
        }
    }
}
